### PR TITLE
Add -XX:ActiveProcessorCount= flag to specify number of CPUs

### PR DESCRIPTION
**Pull request by Ali-Ni**
This change adds the `-XX:ActiveProcessorCount=` flag to specify
the number of CPUs the VM considers to be available. This value
overrides all other forms of CPU detection. The need for this
change is discussed in #1166.

This change depends on eclipse/omr#2673.

Signed-off-by: Mohammad Ali Nikseresht <anikser@ibm.com>

### CODE REVIEW

**Comment by ashu-mehra** [comment]
I am wondering if we should just be using entitled_CPUs for this purpose. Is there any difference between `entitled_CPUs` and `active_CPUs`?

I can see only one difference between Hotspot's `-XX:ActiveProcessorCount` and `-XentitledCPUs`: in hotspots's case it seems whatever value is provided to `-XX:ActiveProcessorCount`, it is used as it is, but in OpenJ9 we use minimum of `-XentitledCPUs` and OMRPORT_CPU_BOUND.
So if we want to use `entitled_CPUs` for this purpose, then this check would need to be removed.

**Comment by Ali-Ni** [comment]
This is what I was thinking as well. The only difference between the two seem to be that `-XX:ActiveProcessorCount` does not have any safeguards. I'm thinking it may be more useful as well - suppose the user wants `-XX:+UseContainerSupport` for the cgroup memory limit, but for whatever reason does not want the CPU quotas to be considered. Overall this flag seems to be more useful for performance tuning, if a bit finicky. 

The question is, what was the reasoning behind adding `-XentitledCPUs`, and considering the new flag, should we remove it now?

**Comment by ashu-mehra** [comment]
From practical point of view I don't see any difference between existing behavior of `-XentitledCPUs` and the behavior expected of `-XX:ActiveProcessorCount`.
So my vote is to just replace `-XentitledCPUs` option name with `-XX:ActiveProcessorCount`.

**Comment by tajila** [comment]
Peter S. suggests deprecating the existing `-XentitledCPUs` option and supporting `-XX:ActiveProcessorCount` with the same behaviour as in hotspot

**Comment by pshipton** [comment]
Note -XentitledCPUs is not a documented option, although available for everyone to see in open source. Not sure what deprecation entails. Perhaps changing the option to print a warning and suggest the new option. And then after at least 3 months, and an OpenJ9 release that includes the change, remove it completely.

**Comment by ashu-mehra** [comment]
@pshipton @DanHeidinga 
Perhaps time to standardize and document the process for deprecating options/APIs? I think we used to have it earlier before we went open.
Eclipse project has a deprecation policy for APIs which is well documented [here](https://wiki.eclipse.org/Eclipse/API_Central/Deprecation_Policy).
OpenJDK has also added more clarity on API deprecation in this [jep](http://openjdk.java.net/jeps/277) and I guess deprecation policy for OpenJ9 specific Java APIs  would follow the same.
But I didn't find any policy for deprecation of VM options.

**Comment by pshipton** [comment]
The option isn't really an "api". The feature wasn't approved or finalized, the option was never documented and was never intended for anybody to use it. We could just remove the option, but since its possible the option leaked out and somebody is using it, I suggested the warning approach, but open to other suggestions.

**Comment by DanHeidinga** [comment]
@ashu-mehra I've opened https://github.com/eclipse/openj9/issues/2212 to track the deprecation policy discussion so we don't side track this PR.

**Comment by Ali-Ni** [comment]
Would entitled CPU functionality be deprecated in OMR as well?

File: runtime/vm/jvminit.c [hunk 1]
```diff
@@ -1610,7 +1610,20 @@ IDATA VMInitStages(J9JavaVM *vm, IDATA stage, void* reserved) {
 				goto _memParseError;
 			}
 
-			/* Set entitled CPUs as early as possible, i.e. as soon as PORT_LIBRARY_GUARANTEED */
+			/* Set active and entitled CPUs as early as possible, i.e. as soon as PORT_LIBRARY_GUARANTEED */
+			if ((argIndex = FIND_AND_CONSUME_ARG(STARTSWITH_MATCH, VMOPT_XXACTIVEPROCESSORCOUNT_EQUALS, NULL)) >= 0) {
```
Comment (DanHeidinga) [inline runtime/vm/jvminit.c]: Can you split this into two lines?  The first assigning to argIndex and the second doing the >= check?

Code is read more often than it's written.  Having each line do one thing helps the reader to easily and quickly see what's happening in the code.

File: runtime/vm/jvminit.c [hunk 2]
```diff
@@ -1610,7 +1610,20 @@ IDATA VMInitStages(J9JavaVM *vm, IDATA stage, void* reserved) {
...
+				UDATA value;
```
Comment (DanHeidinga) [inline runtime/vm/jvminit.c]: nitpick: assign a default value (likely 0)

**Comment by ashu-mehra** [comment]
`entitled_cpus` is also exposed by `ProcessorMXBean` which allows user to get/set it at runtime. Would these APIs also be deprecated?
I am wondering if we can just deprecate `-XentitledCPUs` option but use its functionality for `-XX:ActiveProcessorCount`, i.e. use `j9sysinfo_set_number_entitled_CPUs` for `-XX:ActiveProcessorCount`?

**Comment by pshipton** [comment]
ProcessorMXBean is not properly installed as a MXBean, see [1]. Its just left over code from a failed attempt to add this feature.

[1] jcl/src/jdk.management/share/classes/com/ibm/lang/management/internal/PlatformMBeanProvider.java

**Comment by pshipton** [comment]
> use j9sysinfo_set_number_entitled_CPUs for -XX:ActiveProcessorCount?

I don't think the existing omr code will properly support -XX:ActiveProcessorCount. This allows you to set a number of cpus greater than the physical or bound cpus, but TARGET currently uses the minimum of entitled and bound. If the omr code needs to be modified, there should likely be a new cpu type added rather than trying to change the behavior of the existing ENTITLED type. But this should be discussed at the omr project.

**Comment by SueChaplain** [comment]
The doc:externals label is used, but I don't see an issue created at the openj9-docs repo as per our [labels process](https://github.com/eclipse/openj9/blob/master/doc/processes/labels.md) or [Contributions guide](https://github.com/eclipse/openj9/blob/master/CONTRIBUTING.md).

Please create one here: https://github.com/eclipse/openj9-docs/issues/new?template=new-documentation-change.md

File: jcl/src/jdk.management/share/classes/com/ibm/lang/management/internal/ProcessorMXBeanImpl.java [hunk 3]
```diff
@@ -117,4 +117,30 @@ public int getNumberBoundCPUs() {
...
 	}
+	
+	/**
+	 * @param number
+	 * 			The number of CPUs to specify the process to use. The process
+	 * 			will behave as if <code>number</code> CPUs are available.
+	 * 
+	 * @see #setNumberActiveCPUs(int)
+	 */
+	native void setNumberActiveCPUsImpl(int number);
```
Comment (pshipton) [inline jcl/src/jdk.management/share/classes/com/ibm/lang/management/internal/ProcessorMXBeanImpl.java]: Any reason not to declare as private?

File: runtime/tests/port/si_numcpusTest.c [hunk 4]
```diff
@@ -116,17 +116,45 @@ int j9sysinfo_numcpus_test5 (J9PortLibrary* portLibrary)
...
+	UDATA targetNumberCPUs = 0;
+
+	j9sysinfo_set_number_user_specified_CPUs(J9SYSINFO_NUMCPUS_TEST6_SPECIFIED);
+	targetNumberCPUs = j9sysinfo_get_number_CPUs_by_type(J9PORT_CPU_TARGET);
+
+	outputComment(PORTLIB,"\nSet number of entitled CPUs to %d. Get target number of CPUs.\n", J9SYSINFO_NUMCPUS_TEST6_SPECIFIED);
+	outputComment(PORTLIB,"	Expected: %d\n", J9SYSINFO_NUMCPUS_TEST6_SPECIFIED);
+	outputComment(PORTLIB,"	Result:   %d\n", targetNumberCPUs);
+
+	if (J9SYSINFO_NUMCPUS_TEST6_SPECIFIED != targetNumberCPUs) {
```
Comment (pshipton) [inline runtime/tests/port/si_numcpusTest.c]: There should also be a test that sets active cpus to zero, after setting something else, and confirms the target value goes back to the original value.
Plus a test that confirms a negative value cannot be set.

File: jcl/src/jdk.management/share/classes/com/ibm/lang/management/internal/ProcessorMXBeanImpl.java [hunk 5]
```diff
@@ -117,4 +117,30 @@ public int getNumberBoundCPUs() {
...
+	 * {@inheritDoc}
+	 */
+		@Override
+	public void setNumberActiveCPUs(int number) throws IllegalArgumentException {
+		SecurityManager security = System.getSecurityManager();
+		if (security != null) {
+			security.checkPermission(ManagementPermissionHelper.MPCONTROL);
+		}
+
+		if (0 > number) {
```
[the remark this answers is not in the document]
Comment (Ali-Ni) [reply jcl/src/jdk.management/share/classes/com/ibm/lang/management/internal/ProcessorMXBeanImpl.java]: Double checking with HotSpot, I found that they don't throw an error on negative numbers either, and only consider active cpu's with a valid input. Should we duplicate this behavior or throw an error on all invalid inputs?

[the remark this answers is not in the document]
Comment (pshipton) [reply jcl/src/jdk.management/share/classes/com/ibm/lang/management/internal/ProcessorMXBeanImpl.java]: What is the scenario you tested exactly? Does Hotspot allow setting the active cpus via java code?

[the remark this answers is not in the document]
Comment (Ali-Ni) [reply jcl/src/jdk.management/share/classes/com/ibm/lang/management/internal/ProcessorMXBeanImpl.java]: I tested using the `-XX:ActiveProcessorCount` flag - I really should have asked two separate questions:
1. Should the flag have the exact same behaviour as HotSpot?
2. Should `setNumberActiveCPUs` behave the same as the flag? (I should not have assumed the answer to this be yes)

[the remark this answers is not in the document]
Comment (Ali-Ni) [reply jcl/src/jdk.management/share/classes/com/ibm/lang/management/internal/ProcessorMXBeanImpl.java]: Also, I have not found any java API to set active cpus in HotSpot.

[the remark this answers is not in the document]
Comment (pshipton) [reply jcl/src/jdk.management/share/classes/com/ibm/lang/management/internal/ProcessorMXBeanImpl.java]: It makes sense to me to set zero cpus, which removes the active cpu settings. However the doc and the tests should be updated to cover this.

I suppose there is nothing wrong with setting zero cpus via the command line option, it just doesn't do anything. How does HotSpot handle zero?

[the remark this answers is not in the document]
Comment (Ali-Ni) [reply jcl/src/jdk.management/share/classes/com/ibm/lang/management/internal/ProcessorMXBeanImpl.java]: HotSpot's command line option accepts zero and negative numbers without throwing an error. In this case, it seems to have no effect (`Runtime.availableProcessors()` , at least, is not overridden, and I can't imagine what other course of action the JVM could take).

[the remark this answers is not in the document]
Comment (pshipton) [reply jcl/src/jdk.management/share/classes/com/ibm/lang/management/internal/ProcessorMXBeanImpl.java]: From looking at the code and a simple experiment, OpenJ9 won't accept negative numbers because its parsing for an unsigned value. OpenJ9 already accepts zero and does nothing, which matched the Hotspot behavior. I think the OpenJ9 command line operation is fine.

[the remark this answers is not in the document]
Comment (Ali-Ni) [reply jcl/src/jdk.management/share/classes/com/ibm/lang/management/internal/ProcessorMXBeanImpl.java]: Also, I was under the impression that ProcessorMXBean is not properly installed as an MXBean and is essentially unused code - I added back whatever was already in place (before entitled CPUs was removed) with updated comments and function names since this PR is mostly just undoing #2236 removals. 

Should this just be completely scrapped? If not, is there a way for tests to be written for this API?

[the remark this answers is not in the document]
Comment (pshipton) [reply jcl/src/jdk.management/share/classes/com/ibm/lang/management/internal/ProcessorMXBeanImpl.java]: The code can be called directly (in Java 8), as shown by the testing, its just not a proper MXBean. I think updating it is correct. We could expose it as a MXBean in a subsequent change.

[the remark this answers is not in the document]
Comment (pshipton) [reply jcl/src/jdk.management/share/classes/com/ibm/lang/management/internal/ProcessorMXBeanImpl.java]: Sorry, I guess there is no testing for this class, but the code can be called directly by creating an instance of ProcessorMXBeanImpl

[the remark this answers is not in the document]
Comment (pshipton) [reply jcl/src/jdk.management/share/classes/com/ibm/lang/management/internal/ProcessorMXBeanImpl.java]: To be clear, I'm not asking for tests for ProcessorMXBeanImpl in order to merge this.

[the remark this answers is not in the document]
Comment (Ali-Ni) [reply jcl/src/jdk.management/share/classes/com/ibm/lang/management/internal/ProcessorMXBeanImpl.java]: OK, understood. I was confused earlier about how this API was exposed and what the implications of that were - thanks for the clarification

File: runtime/tests/port/si_numcpusTest.c (hunk 4 above)
Comment (Ali-Ni) [reply runtime/tests/port/si_numcpusTest.c]: Right now, a negative value can technically be set (omr does not throw an error in `j9sysinfo_set_number_user_specified_CPUs`), but it would have the same effect as setting 0 (active CPUs are then not considered).

But since the command line option cannot parse negative numbers, and negative numbers are not accepted by the ProcessorMXBean function for consistency, a negative value effectively cannot be set (there just isn't a test to be written here). I think this is fine.

**Comment by Ali-Ni** [comment]
There is no CPU type associated with the active/userSpecified CPUs - when set, it just overrides `J9PORT_CPU_TARGET`, which is otherwise equal to `J9PORT_CPU_BOUND`. Since number of bound and target CPUs is there, there is no information missing from the javacore.

File: runtime/jcl/common/mgmtprocessor.c [hunk 6]
```diff
@@ -45,3 +45,17 @@ Java_com_ibm_lang_management_internal_ProcessorMXBeanImpl_getNumberCPUsImpl(JNIE
...
 }
+
+/**
+ * Makes the Port Library function j9sysinfo_set_number_user_specified_CPUs(number) available
+ * for JCL.
+ *
+ * @param[in] number Number of user-specified active CPUs (non-negative).
+ */
+void JNICALL
+Java_com_ibm_lang_management_internal_ProcessorMXBeanImpl_setNumberActiveCPUsImpl(JNIEnv *env, jobject o, jint number)
```
[the remark this answers is not in the document]
Comment (Ali-Ni) [reply runtime/jcl/common/mgmtprocessor.c]: Added

File: runtime/tests/port/si_numcpusTest.c (hunk 4 above)
Comment (pshipton) [reply runtime/tests/port/si_numcpusTest.c]: Yes I agree. I didn't realize at first that omrsysinfo_get_number_CPUs_by_type takes an unsigned value.

**Comment by pshipton** [comment]
Hmm, it would be nicer to show the value explicitly rather than derive it, as it could cause confusion. How about changing "Target CPUs:" to "Target/Active CPUs:" to give users a hint.

File: runtime/tests/port/si_numcpusTest.c [hunk 7]
```diff
@@ -116,17 +116,76 @@ int j9sysinfo_numcpus_test5 (J9PortLibrary* portLibrary)
...
+int j9sysinfo_numcpus_test5 (J9PortLibrary* portLibrary)
+{
+#define J9SYSINFO_NUMCPUS_TEST5_SPECIFIED 3
+	PORT_ACCESS_FROM_PORT(portLibrary);
+	const char* testName = "j9sysinfo_numcpus_test5";
+	UDATA targetNumberCPUsOriginal = 0;
+	UDATA targetNumberCPUs = 0;
+
+	targetNumberCPUsOriginal = j9sysinfo_get_number_CPUs_by_type(J9PORT_CPU_TARGET);
+	j9sysinfo_set_number_user_specified_CPUs(J9SYSINFO_NUMCPUS_TEST5_SPECIFIED);
```
Comment (pshipton) [inline runtime/tests/port/si_numcpusTest.c]: The test isn't resetting the value to zero.

**Comment by Ali-Ni** [comment]
Alternatively, we could add something like "(overridden by Active CPUs)" if target is being overridden - something like this:
```
UDATA bound = j9sysinfo_get_number_CPUs_by_type(J9PORT_CPU_BOUND);
UDATA target = j9sysinfo_get_number_CPUs_by_type(J9PORT_CPU_TARGET);

_OutputStream.writeCharacters(
		"2CIBOUNDCPU   Bound CPUs: ");
_OutputStream.writeInteger(bound, "%i\n");
_OutputStream.writeCharacters(
		"2CITARGETCPU   Target CPUs");
if (bound != target) {
	_OutputStream.writeCharacters(" (overridden by Active CPUs)");
} 
_OutputStream.writeInteger(target, ": %i\n");
```
What are your thoughts?

**Comment by pshipton** [comment]
Looks good. It may affect the doc, please mention this change in https://github.com/eclipse/openj9-docs/issues/60

**Comment by pshipton** [comment]
It may be unexpected to any tool parsing the file. On second thought, you can use the same logic to add a new line for Active CPUs - 2CIACTIVECPU.

File: runtime/rasdump/javadump.cpp [hunk 8]
```diff
@@ -5120,10 +5123,17 @@ JavaCoreDumpWriter::writeCPUinfo(void)
 	_OutputStream.writeInteger(j9sysinfo_get_number_CPUs_by_type(J9PORT_CPU_ONLINE), "%i\n");
 	_OutputStream.writeCharacters(
 			"2CIBOUNDCPU    Bound CPUs: ");
-	_OutputStream.writeInteger(j9sysinfo_get_number_CPUs_by_type(J9PORT_CPU_BOUND), "%i\n");
+	_OutputStream.writeInteger(bound, "%i\n");
 	_OutputStream.writeCharacters(
 			"2CITARGETCPU   Target CPUs: ");
-	_OutputStream.writeInteger(j9sysinfo_get_number_CPUs_by_type(J9PORT_CPU_TARGET), "%i\n");
+	_OutputStream.writeInteger(target, "%i\n");
+	_OutputStream.writeCharacters(
```
Comment (pshipton) [inline runtime/rasdump/javadump.cpp]: I think 2CIACTIVECPU should go before 2CITARGETCPU, so 2CITARGETCPU is last, since 2CITARGETCPU is what is being used to control number of threads created.

File: runtime/rasdump/javadump.cpp [hunk 9]
```diff
@@ -5120,10 +5123,18 @@ JavaCoreDumpWriter::writeCPUinfo(void)
...
 			"2CIBOUNDCPU    Bound CPUs: ");
-	_OutputStream.writeInteger(j9sysinfo_get_number_CPUs_by_type(J9PORT_CPU_BOUND), "%i\n");
+	_OutputStream.writeInteger(bound, "%i\n");
+	_OutputStream.writeCharacters(
+			"2CIACTIVECPU   Active CPUs: ");
+	if (bound != target) {
+		_OutputStream.writeInteger(target, "%i\n");
+	} else {
+		/* target is not being overridden by active CPUs, so print 0 */
+		_OutputStream.writeInteger(0, "%i\n");
```
Comment (pshipton) [inline runtime/rasdump/javadump.cpp]: Seems "0\n" would work just as well.

**Comment by pshipton** [comment]
If you make an update, please make a comment about it so I get an email notification about the change.

File: runtime/rasdump/javadump.cpp (hunk 9 above)
Comment (Ali-Ni) [reply runtime/rasdump/javadump.cpp]: Right, haha. Done.

**Comment by pshipton** [comment]
jenkins test all zlinux jdk8

**Comment by Ali-Ni** [comment]
Test 5 would not run properly as entitled CPUs was being set earlier in test 4. Switched the test order to resolve this.

**Comment by pshipton** [comment]
jenkins compile win jdk8

**Comment by pshipton** [comment]
> Test 5 would not run properly as active CPUs was being set earlier in test 4. Switched the test order to resolve this.

Really the test should set the value back to zero at the end, to avoid affecting other testing.

**Comment by SueChaplain** [comment]
@pshipton @Ali-Ni Which OpenJ9 release gets this change please?
